File: api/schemas/card_terms.py
"""Phase 7b — credit-card terms schemas (B3) + analysis response (B4).

Rates are 0–1 fractions everywhere (18% → 0.18), matching `debts.interest_rate`.
`CardTermsExtraction` mirrors `DebtTermsExtraction`: every field Optional except
`confidence` so the model can honestly leave unknowns null — the form asks for
whatever is missing. "LLM extracts; rules decide": the deterministic
PUT /accounts/{id}/card-terms is the only write path.
"""
import uuid
from datetime import date, datetime
from decimal import Decimal
from typing import Literal, Optional

from pydantic import BaseModel, Field


class CardTermsCreate(BaseModel):
    model_config = {"extra": "forbid"}

    annual_interest_rate: Decimal = Field(..., ge=0, le=1)
    cash_advance_rate: Optional[Decimal] = Field(None, ge=0, le=1)
    minimum_payment_pct: Decimal = Field(..., gt=0, le=1)
    minimum_payment_floor: Optional[Decimal] = Field(None, gt=0)
    credit_limit: Optional[Decimal] = Field(None, gt=0)
    statement_day: Optional[int] = Field(None, ge=1, le=31)
    payment_due_day: int = Field(..., ge=1, le=31)
    # Recurring-payment preference: 'minimum' (default) or 'full' (pago de
    # contado). Selects which live figure the upcoming-payment projection
    # surfaces; budget/affordability/reservation stay on the minimum.
    payment_mode: Literal["minimum", "full"] = "minimum"
    # B5 obligation attachment: the card's minimum reserves inside this
    # envelope while unpaid (validated active + same-user in the router).
    envelope_id: Optional[uuid.UUID] = None
    notes: Optional[str] = Field(None, max_length=500)


class CardTermsUpdate(BaseModel):
    model_config = {"extra": "forbid"}

    annual_interest_rate: Optional[Decimal] = Field(None, ge=0, le=1)
    cash_advance_rate: Optional[Decimal] = Field(None, ge=0, le=1)
    minimum_payment_pct: Optional[Decimal] = Field(None, gt=0, le=1)
    minimum_payment_floor: Optional[Decimal] = Field(None, gt=0)
    credit_limit: Optional[Decimal] = Field(None, gt=0)
    statement_day: Optional[int] = Field(None, ge=1, le=31)
    payment_due_day: Optional[int] = Field(None, ge=1, le=31)
    payment_mode: Optional[Literal["minimum", "full"]] = None
    notes: Optional[str] = Field(None, max_length=500)


class CardTermsResponse(BaseModel):
    id: uuid.UUID
    user_id: uuid.UUID
    account_id: uuid.UUID
    annual_interest_rate: Decimal
    cash_advance_rate: Optional[Decimal]
    minimum_payment_pct: Decimal
    minimum_payment_floor: Optional[Decimal]
    credit_limit: Optional[Decimal]
    statement_day: Optional[int]
    payment_due_day: int
    payment_mode: str
    envelope_id: Optional[uuid.UUID]
    notes: Optional[str]
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class CardTermsExtraction(BaseModel):
    """Phase 7b B3 — card terms parsed from an uploaded contract (or
    statement) PDF.

    Returned by `POST /accounts/parse-card-document` to PRE-FILL the native
    card form; it does NOT create anything. Rates are 0–1 fractions.

    Dual-currency CR cards: the base fields are the COLONES terms; the
    `*_usd` fields are filled only when the document lists separate dollar
    terms — their presence is the signal that the card operates in both
    currencies (the form then creates one credit account per currency).
    """

    issuer: Optional[str] = None
    credit_limit: Optional[float] = Field(None, gt=0)
    statement_balance: Optional[float] = Field(None, ge=0)
    annual_interest_rate: Optional[float] = Field(None, ge=0, le=1)
    cash_advance_rate: Optional[float] = Field(None, ge=0, le=1)
    annual_interest_rate_usd: Optional[float] = Field(None, ge=0, le=1)
    credit_limit_usd: Optional[float] = Field(None, gt=0)
    statement_balance_usd: Optional[float] = Field(None, ge=0)
    minimum_payment_pct: Optional[float] = Field(None, gt=0, le=1)
    minimum_payment_amount: Optional[float] = Field(None, gt=0)
    statement_day: Optional[int] = Field(None, ge=1, le=31)
    payment_due_day: Optional[int] = Field(None, ge=1, le=31)
    currency: Optional[str] = None
    confidence: float = Field(..., ge=0, le=1)


# ── B4: deterministic card analysis (engine output over the live balance) ─────


class CardPaymentStrategy(BaseModel):
    label: str
    monthly_payment: Decimal
    months: Optional[int]
    total_interest: Decimal
    interest_saved_vs_minimum: Optional[Decimal]


class CardAnalysisResponse(BaseModel):
    account_id: uuid.UUID
    currency: str
    as_of: date
    balance_owed: Decimal
    credit_limit: Optional[Decimal]
    available_credit: Optional[Decimal]
    minimum_payment: Decimal
    monthly_interest_cost: Decimal
    months_to_payoff_minimum: Optional[int]
    total_interest_minimum: Optional[Decimal]
    total_paid_minimum: Optional[Decimal]
    never_pays_off: bool
    strategies: list[CardPaymentStrategy]

File: mobile/src/api/cardTerms.ts
/**
 * Phase 7b B3/B4 — credit-card terms + analysis.
 *
 * Terms are ACCOUNT PARAMETERS (1:1 with the credit account), not a Debt.
 * Rates travel as 0–1 fractions (the UI shows percent and converts).
 * The analysis is computed server-side by the deterministic revolving engine
 * over the LIVE balance — the app never reimplements the interest math.
 */
import { api, LLM_UPLOAD_TIMEOUT_MS } from "./client";

export interface CardTermsPayload {
  annual_interest_rate: string; // 0–1 fraction
  cash_advance_rate?: string | null;
  minimum_payment_pct: string; // 0–1 fraction
  minimum_payment_floor?: string | null;
  credit_limit?: string | null;
  statement_day?: number | null;
  payment_due_day: number;
  /** 'minimum' (must-pay floor) or 'full' (pago de contado — full balance).
   * Selects which live amount the upcoming-payment reminder + agent surface. */
  payment_mode?: "minimum" | "full";
  /** B5 attachment — omit to keep the current envelope; null detaches. */
  envelope_id?: string | null;
  notes?: string | null;
}

export interface CardTermsResponse {
  id: string;
  user_id: string;
  account_id: string;
  annual_interest_rate: string;
  cash_advance_rate: string | null;
  minimum_payment_pct: string;
  minimum_payment_floor: string | null;
  credit_limit: string | null;
  statement_day: number | null;
  payment_due_day: number;
  payment_mode: "minimum" | "full";
  envelope_id: string | null;
  notes: string | null;
  created_at: string;
  updated_at: string;
}

export interface CardTermsExtraction {
  issuer: string | null;
  credit_limit: number | null;
  statement_balance: number | null;
  annual_interest_rate: number | null; // 0–1 fraction (colones terms)
  cash_advance_rate: number | null;
  // Dual-currency CR cards: present only when the contract lists separate
  // dollar terms — their presence is the "this card is ₡ + $" signal.
  annual_interest_rate_usd: number | null;
  credit_limit_usd: number | null;
  statement_balance_usd: number | null;
  minimum_payment_pct: number | null; // 0–1 fraction
  minimum_payment_amount: number | null;
  statement_day: number | null;
  payment_due_day: number | null;
  currency: string | null;
  confidence: number;
}

export interface CardPaymentStrategy {
  label: string;
  monthly_payment: string;
  months: number | null;
  total_interest: string;
  interest_saved_vs_minimum: string | null;
}

export interface CardAnalysisResponse {
  account_id: string;
  currency: string;
  as_of: string;
  balance_owed: string;
  credit_limit: string | null;
  available_credit: string | null;
  minimum_payment: string;
  monthly_interest_cost: string;
  months_to_payoff_minimum: number | null;
  total_interest_minimum: string | null;
  total_paid_minimum: string | null;
  never_pays_off: boolean;
  strategies: CardPaymentStrategy[];
}

export async function fetchCardTerms(
  accountId: string,
): Promise<CardTermsResponse | null> {
  try {
    const { data } = await api.get<CardTermsResponse>(
      `/accounts/${accountId}/card-terms`,
    );
    return data;
  } catch (err) {
    const status = (err as { response?: { status?: number } }).response?.status;
    if (status === 404) return null; // no terms yet — the UI offers "Agregar"
    throw err;
  }
}

export async function upsertCardTerms(
  accountId: string,
  payload: CardTermsPayload,
): Promise<CardTermsResponse> {
  const { data } = await api.put<CardTermsResponse>(
    `/accounts/${accountId}/card-terms`,
    payload,
  );
  return data;
}

export async function parseCardDocument(
  uri: string,
): Promise<CardTermsExtraction> {
  const form = new FormData();
  const filename = uri.split("/").pop() ?? "estado.pdf";
  form.append("file", {
    uri,
    type: "application/pdf",
    name: filename,
  } as unknown as Blob);
  const { data } = await api.post<CardTermsExtraction>(
    "/accounts/parse-card-document",
    form,
    {
      headers: { "Content-Type": "multipart/form-data" },
      timeout: LLM_UPLOAD_TIMEOUT_MS,
    },
  );
  return data;
}

export async function fetchCardAnalysis(
  accountId: string,
): Promise<CardAnalysisResponse> {
  const { data } = await api.get<CardAnalysisResponse>(
    `/accounts/${accountId}/card-analysis`,
  );
  return data;
}
